docs/04-reference/winglang-spec.md

```diff
@@ -869,30 +869,223 @@ let f = (arg1: num, var arg2: num) => {
+
+```js
+let parse_int = (x: str, radix: num?, opts?: ParseOpts): num { /* ... */ };
+// or
+let parse_int = (x: str, radix: num = 10, opts: ParseOpts = ParseOpts {}): num { /* ... */ };
+```
+
+##### 1.7.1.5 Declaration in function return types
+
+If a function returns an optional type, use the `return nil;` statement to indicate that the value is
+not defined.
+
+> This is currently the only scenario in the language where `nil` is allowed.
+
+```js
+struct Name { first: str, last: str };
+
+let try_parse_name = (full_name: str): Name? => {
+  let parts = full_name.split(" ");
+  if parts.len < 2 {
+    return nil;
+  }
+
+  return Name { first: parts.at(0), last: parts.at(1) };
+}
+```
+
+#### 1.7.2 Testing using `x?`
+
+Object of type `str?` cannot be used within `==` or `!=` expressions because we don't have a null
+value. Instead, use the the `?` operator to get a `bool` that indicates if the optional is defined
+or not:
+
+```js
+let is_address_defined = my_person.address?; // type is `bool`
+
+// or within a condition
+if my_person.address? {
+  print("address is defined but i do not care what it is");
+}
+
+// can be negated
+if !my_person.address? {
+  print("address is not defined");
+}
+```
+
+#### 1.7.3 Unwrapping using `if let` and `let else`
+
+The `if let` statement can be used to test if an optional is defined an *unwrap* it into a
+non-optional variable defined inside the block:
+
+```js
+if let address = my_person.address {
+  print(address.len);
+  print(address); // address is type `str`
+}
+```
+
+> NOTE: `if let` is not the same as `if`. For example, we currently don't support specifying multiple conditions, or unwrapping multiple optionals. This is something we might consider in the future.
+
+The `let .. else` statement can be used to conditionally execute a block if an optional
+does not have a value:
+
+```js
+let address = my_person.address else {
+  throw("address is required");
+};
+
+print("address is ${address}");
+```
+
+#### 1.7.4 Unwrapping or default value using `??`
+
+The `??` operator can be used to unwrap or provide a default value. This returns a value of `T` that can safely be used.
+
+```js
+let address: str = my_person.address ?? "Planet Earth";
+```
+
+#### 1.7.5 Optional chaining using `?.`
+
+The `?.` syntax can be used for optional chaining. Optional chaining returns a value of type `T?` which must be unwrapped in order to be used.
+
+```js
+if let ip_address = options.networking?.ip_address {
+  print("the ip address is defined and it is: ${ip_address}");
+}
+```
+
+#### 1.7.6 Roadmap
+
+In the future we will consider the following additional sugar syntax:
+
+* `x ?? throw("message")` to unwrap `x` or throw if `x` is not defined.
+* `x ??= value` returns `x` or assigns a value to it and returns it to support lazy evaluation/memoization (inspired by [Nullish coalescing assignment](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_assignment)).
```

Comment:
Other things I'd like:
* multiple `if lets` in a single expression. Can be initially supported only with `&&` operator:
```wing
if let x = x && let y = y { .. }
```
* Support `&&` operator in `if let`s:
```wing
if its_raining && let x = x { .. }
```
* Support implicit naming of unwrapped variable for `if let`s:
```wing
if let x { print("unwrapped x: ${x}"); }
```

Reply:
> * Support implicit naming of unwrapped variable for `if let`s:
> 
> ```
> if let x { print("unwrapped x: ${x}"); }
> ```

I think this is handled by the other syntax? I could be mistaken

```js
if let address = maybe_address {
  print("unwrapped address: ${address}");
}
```

Reply:
Imaging this case:
```wing
if let address = address { print("${address}"}; }
// We can have sugar here so this will be the same as:
if let address { print("${address}"}; }
// Or we can make the syntax a bit different to make this case clearer
if unwrap address { print("${address}"}; }
```

Reply:
Leaving as is for now